src/tokenRefresh.ts

```diff
@@ -1,41 +1,41 @@
 import apiClient from "./apiClient"
-import { Service, postData } from "./base"
-
-const refreshToken = async () => {
-  const { refresh_token, access_token } = await postData(
-    Service.TOKEN_REFRESH,
-    "refresh-token",
-    { token: apiClient.tokens.refreshToken }
-  )
-
-  apiClient.setTokens({
-    accessToken: access_token,
-    refreshToken: refresh_token,
-  })
+
+const invokeTokenRefresh = async () => {
+  const { accessToken } = await apiClient.refreshToken()
+
+  apiClient.setAccessToken(accessToken)
   if (apiClient.handlers.onAccessTokenUpdate) {
-    apiClient.handlers.onAccessTokenUpdate(apiClient.tokens.accessToken)
+    apiClient.handlers.onAccessTokenUpdate(apiClient.accessToken)
   }
 }
 
+let pendingRefreshRequest = false
+const waitForPendingRefresh = async (apiCall) => {
+  return new Promise((resolve) => {
+    setTimeout(async () => {
+      return pendingRefreshRequest
+        ? resolve(waitForPendingRefresh(apiCall))
+        : resolve(apiCall())
+    }, 100)
```

Comment:
can there be a case where this causes an endless loop or a very long running one that we'd want to time out at some point?

Reply:
that said - you're in dangerous territory here. these things are hard to test and hard to detect in production. if you have a simpler approach, I'd prefer that over potential endless loop